File: ios/CodexMeterCore/Sources/CodexMeterCore/DashboardSections.swift
import Foundation

/// Stable keys and merge rules for movable dashboard sections.
public enum DashboardSections {
    public static let fiveHour = "five_hour"
    public static let weekly = "weekly"
    public static let monthly = "monthly"
    public static let usageCredits = "usage_credits"
    public static let usageHistory = "usage_history"
    public static let resetCredits = "reset_credits"

    private static let limitPrefix = "limit:"

    /// A stable key for an additional limit, preferring the API id over display names.
    public static func limitKey(_ limit: UsageLimit) -> String {
        let identity = [limit.id, limit.name, limit.meteredFeature, "additional"]
            .first { !$0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }!
        let normalized = identity
            .lowercased(with: Locale(identifier: "en_US_POSIX"))
            .replacingOccurrences(of: ",", with: "_")
        return limitPrefix + normalized
    }

    public static func isLimitKey(_ key: String) -> Bool {
        key.hasPrefix(limitPrefix)
    }

    /// Default order: standard windows, monthly, detected additional limits, credits, history, resets.
    public static func defaultOrder(additionalLimits: [UsageLimit]) -> [String] {
        var result = [fiveHour, weekly, monthly]
        for limit in additionalLimits {
            let key = limitKey(limit)
            if !result.contains(key) {
                result.append(key)
            }
        }
        result.append(contentsOf: [usageCredits, usageHistory, resetCredits])
        return result
    }

    /// Applies a saved order to the currently available section keys.
    ///
    /// Unavailable saved keys are removed. Newly available keys are inserted after the nearest
    /// preceding key in the natural `available` order instead of being dumped at the end.
    public static func resolveOrder(saved: [String], available: [String]) -> [String] {
        let availableKeys = deduplicating(available)
        var result = deduplicating(saved).filter(availableKeys.contains)

        for (index, key) in availableKeys.enumerated() where !result.contains(key) {
            var insertionIndex = 0
            for precedingKey in availableKeys.prefix(index) {
                if let position = result.firstIndex(of: precedingKey) {
                    insertionIndex = max(insertionIndex, result.index(after: position))
                }
            }
            result.insert(key, at: insertionIndex)
        }
        return result
    }

    public static func resolveOrder(savedCSV: String, available: [String]) -> [String] {
        resolveOrder(saved: parseCSV(savedCSV), available: available)
    }

    public static func isHidden(_ hiddenCSV: String, key: String) -> Bool {
        parseCSV(hiddenCSV).contains(key)
    }

    /// Adds or removes a key from a comma-separated hidden-section setting.
    public static func settingHidden(_ hiddenCSV: String, key: String, hidden: Bool) -> String {
        let key = key.trimmingCharacters(in: .whitespacesAndNewlines)
        var keys = deduplicating(parseCSV(hiddenCSV))
        guard !key.isEmpty else {
            return serialize(keys)
        }
        if hidden {
            if !keys.contains(key) {
                keys.append(key)
            }
        } else {
            keys.removeAll { $0 == key }
        }
        return serialize(keys)
    }

    public static func parseCSV(_ csv: String) -> [String] {
        csv.split(separator: ",")
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
    }

    public static func serialize(_ keys: [String]) -> String {
        keys
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
            .joined(separator: ",")
    }

    private static func deduplicating(_ keys: [String]) -> [String] {
        var seen = Set<String>()
        return keys.filter { seen.insert($0).inserted }
    }
}

File: android/shared/src/main/java/dev/scorpion7slayer/modelsmeter/DashboardSections.java
package dev.scorpion7slayer.modelsmeter;

import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Locale;

/**
 * Stable keys and ordering rules for the movable dashboard sections: the built-in 5-hour and
 * weekly windows, every automatically detected additional limit (for example
 * GPT-5.3-Codex-Spark), the usage-credit balance, and the local usage-history charts. The saved
 * order is a comma-separated key list; unknown saved keys are dropped and newly detected
 * sections slot into their default position, so accounts gaining or losing model-specific
 * limits never lose their arrangement. A separate comma-separated hidden-key list tracks the
 * sections a user switched off from the edit screen.
 */
public final class DashboardSections {
    public static final String FIVE_HOUR = "five_hour";
    public static final String WEEKLY = "weekly";
    public static final String MONTHLY = "monthly";
    public static final String USAGE_CREDITS = "usage_credits";
    public static final String USAGE_HISTORY = "usage_history";
    public static final String RESET_CREDITS = "reset_credits";
    public static final String LATEST_MODELS = "latest_models";
    private static final String LIMIT_PREFIX = "limit:";

    private DashboardSections() {
    }

    /** Stable key for an additional limit, preferring the API id over display names. */
    public static String limitKey(UsageLimit limit) {
        String identity = limit.id;
        if (identity.isEmpty()) {
            identity = limit.name;
        }
        if (identity.isEmpty()) {
            identity = limit.meteredFeature;
        }
        if (identity.isEmpty()) {
            identity = "additional";
        }
        return LIMIT_PREFIX + identity.toLowerCase(Locale.ROOT).replace(',', '_');
    }

    public static boolean isLimitKey(String key) {
        return key != null && key.startsWith(LIMIT_PREFIX);
    }

    /**
     * Default order: 5-hour, weekly, monthly, detected additional limits, usage credits,
     * history, reset credits.
     */
    public static List<String> defaultOrder(List<UsageLimit> additionalLimits) {
        List<String> order = new ArrayList<>();
        order.add(LATEST_MODELS);
        order.add(FIVE_HOUR);
        order.add(WEEKLY);
        order.add(MONTHLY);
        if (additionalLimits != null) {
            for (UsageLimit limit : additionalLimits) {
                if (limit != null && !order.contains(limitKey(limit))) {
                    order.add(limitKey(limit));
                }
            }
        }
        order.add(USAGE_CREDITS);
        order.add(USAGE_HISTORY);
        order.add(RESET_CREDITS);

        return order;
    }

    /** Whether a section key is present in the saved hidden-key CSV. */
    public static boolean isHidden(String hiddenCsv, String key) {
        return key != null && parseCsv(hiddenCsv).contains(key);
    }

    /** Adds or removes a key from the hidden-key CSV, deduplicating existing entries. */
    public static String setHidden(String hiddenCsv, String key, boolean hidden) {
        if (key == null || key.trim().isEmpty()) {
            return serialize(parseCsv(hiddenCsv));
        }
        List<String> keys = new ArrayList<>(new LinkedHashSet<>(parseCsv(hiddenCsv)));
        String trimmed = key.trim();
        if (hidden) {
            if (!keys.contains(trimmed)) {
                keys.add(trimmed);
            }
        } else {
            keys.remove(trimmed);
        }
        return serialize(keys);
    }

    /**
     * Applies a saved order to the currently available section keys. Saved keys that are no
     * longer available are ignored; available keys missing from the saved order are inserted
     * right after the nearest preceding available key that survived, so new sections appear in
     * their natural position instead of being dumped at the end.
     */
    public static List<String> resolveOrder(String savedCsv, List<String> available) {
        List<String> availableKeys = new ArrayList<>(new LinkedHashSet<>(available));
        List<String> result = new ArrayList<>();
        for (String key : parseCsv(savedCsv)) {
            if (availableKeys.contains(key) && !result.contains(key)) {
                result.add(key);
            }
        }
        for (int i = 0; i < availableKeys.size(); i++) {
            String key = availableKeys.get(i);
            if (result.contains(key)) {
                continue;
            }
            int insertAt = 0;
            for (int previous = 0; previous < i; previous++) {
                int position = result.indexOf(availableKeys.get(previous));
                if (position >= 0) {
                    insertAt = Math.max(insertAt, position + 1);
                }
            }
            result.add(insertAt, key);
        }
        return result;
    }

    public static String serialize(List<String> order) {
        StringBuilder csv = new StringBuilder();
        for (String key : order) {
            if (key == null || key.trim().isEmpty()) {
                continue;
            }
            if (csv.length() > 0) {
                csv.append(',');
            }
            csv.append(key.trim());
        }
        return csv.toString();
    }

    private static List<String> parseCsv(String csv) {
        List<String> keys = new ArrayList<>();
        if (csv == null || csv.trim().isEmpty()) {
            return keys;
        }
        for (String part : csv.split(",")) {
            String key = part.trim();
            if (!key.isEmpty()) {
                keys.add(key);
            }
        }
        return keys;
    }
}
